MuonAnalysis/MuonAssociators/interface/PropagateToMuonInterface.h

```diff
@@ -0,0 +1,79 @@
+
+  \author   Giovanni Petrucciani
+*/
+
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "DataFormats/GeometrySurface/interface/BoundCylinder.h"
+#include "DataFormats/GeometrySurface/interface/BoundDisk.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "RecoMuon/DetLayers/interface/MuonDetLayerGeometry.h"
+#include "TrackingTools/GeomPropagators/interface/Propagator.h"
+#include "TrackingTools/TrajectoryState/interface/TrajectoryStateOnSurface.h"
+#include "SimDataFormats/Track/interface/SimTrackContainer.h"
+#include "SimDataFormats/Vertex/interface/SimVertexContainer.h"
+#include "MuonAnalysis/MuonAssociators/interface/PropagateToMuonCore.h"
+#include "MuonAnalysis/MuonAssociators/interface/trackStateEnums.h"
+
+class IdealMagneticFieldRecord;
+class TrackingComponentsRecord;
+class MuonRecoGeometryRecord;
+
+class PropagateToMuonInterface {
+public:
+  explicit PropagateToMuonInterface(const edm::ParameterSet &iConfig, edm::ConsumesCollector);
+  ~PropagateToMuonInterface();
+
+  /// Call this method at the beginning of each run, to initialize geometry,
+  /// magnetic field and propagators
+  PropagateToMuonCore init(const edm::EventSetup &iSetup) const;
+  static void fillPSetDescription(edm::ParameterSetDescription &desc) {
+    desc.add<bool>("useSimpleGeometry", true);
+    desc.add<bool>("useStation2", true);
+    desc.add<std::string>("useTrack", "tracker");
+    desc.add<std::string>("useState", "atVertex");
+    desc.add<edm::ESInputTag>("propagatorAlong", edm::ESInputTag("", "hltESPSteppingHelixPropagatorAlong"));
+    desc.add<edm::ESInputTag>("propagatorAny", edm::ESInputTag("", "SteppingHelixPropagatorAny"));
+    desc.add<edm::ESInputTag>("propagatorOpposite", edm::ESInputTag("", "hltESPSteppingHelixPropagatorOpposite"));
+  }
+
+private:
+  /// Use simplified geometry (cylinders and disks, not individual chambers)
+  bool useSimpleGeometry_;
+
+  /// Propagate to MB2 (default) instead of MB1
+  bool useMB2_;
+
+  /// Fallback to ME1 if propagation to ME2 fails
+  bool fallbackToME1_;
+
+  /// Labels for input collections
+  WhichTrack whichTrack_;
+  WhichState whichState_;
+
+  /// for cosmics, some things change: the along-opposite is not in-out, nor the innermost/outermost states are in-out really
+  bool cosmicPropagation_;
+
+  bool useMB2InOverlap_;
+
+  edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> magfieldToken_;
+  edm::ESGetToken<Propagator, TrackingComponentsRecord> propagatorToken_, propagatorAnyToken_, propagatorOppositeToken_;
+  edm::ESGetToken<MuonDetLayerGeometry, MuonRecoGeometryRecord> muonGeometryToken_;
```

Comment:
Can any of these become `const`?

Reply:
All except the `whichTrack` and `whichState`, for which some logic has to be done in the constructor.